### PR TITLE
Create task page

### LINKED ISSUE #17: APP - Create Task Page

### PR DESCRIPTION
**Pull request by henneboy**
Closes #17 

The page that was added:
image
image
image
image
image

### CODE REVIEW

File: frontend/app/src/main/java/dk/scheduling/schedulingfrontend/sharedcomponents/DateRangePicker.kt [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+                                closeDialog()
+                            },
+                            enabled = state.selectedEndDateMillis != null,
+                        ) {
+                            Text(text = "Save")
+                        }
+                    }
+                    DateRangePicker(
+                        state = state,
+                        // modifier = Modifier.fillMaxSize()
```
Comment (henneboy) [inline frontend/app/src/main/java/dk/scheduling/schedulingfrontend/sharedcomponents/DateRangePicker.kt:106]:
```suggestion
```

File: frontend/app/src/main/java/dk/scheduling/schedulingfrontend/sharedcomponents/DateRangePicker.kt [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+                        selectableDates =
+                            object : SelectableDates {
+                                override fun isSelectableDate(utcTimeMillis: Long): Boolean {
+                                    return utcTimeMillis >= System.currentTimeMillis() - 86400000
+                                }
+                            },
+                    )
+
+                Column(
+                    // modifier = Modifier.fillMaxSize(),
```
Comment (henneboy) [inline frontend/app/src/main/java/dk/scheduling/schedulingfrontend/sharedcomponents/DateRangePicker.kt:78]:
```suggestion
```

File: frontend/app/src/main/java/dk/scheduling/schedulingfrontend/sharedcomponents/DateRangePicker.kt [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+@OptIn(ExperimentalMaterial3Api::class)
+@Composable
+fun StandardDateRangePicker(
+    closeDialog: () -> Unit,
+    passingDate: (LongRange) -> Unit,
+    openDialog: Boolean,
+) {
+    if (openDialog) {
+        BasicAlertDialog(
+            // modifier = Modifier.fillMaxSize(),
```
Comment (henneboy) [inline frontend/app/src/main/java/dk/scheduling/schedulingfrontend/sharedcomponents/DateRangePicker.kt:48]:
```suggestion
```

File: frontend/app/src/main/java/dk/scheduling/schedulingfrontend/sharedcomponents/TimePicker.kt [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+import dk.scheduling.schedulingfrontend.ui.theme.SchedulingFrontendTheme
+
+@OptIn(ExperimentalMaterial3Api::class)
+@Composable
+fun StandardTimePickerDialog(
+    closeDialog: () -> Unit,
+    passValue: (TimePickerState) -> Unit,
+    openDialog: Boolean,
+) {
+    // val state = rememberTimePickerState()
```
Comment (henneboy) [inline frontend/app/src/main/java/dk/scheduling/schedulingfrontend/sharedcomponents/TimePicker.kt:42]:
```suggestion
```

File: frontend/app/src/main/java/dk/scheduling/schedulingfrontend/sharedcomponents/TimePicker.kt [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+    // val showingPicker = remember { mutableStateOf(true) }
```
Comment (henneboy) [inline frontend/app/src/main/java/dk/scheduling/schedulingfrontend/sharedcomponents/TimePicker.kt:43]:
```suggestion
```

File: frontend/app/src/main/java/dk/scheduling/schedulingfrontend/sharedcomponents/TimePicker.kt [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+    if (openDialog) {
+        val showingPicker = remember { mutableStateOf(true) }
+        BasicAlertDialog(
+            // modifier = Modifier.fillMaxSize(),
```
Comment (henneboy) [inline frontend/app/src/main/java/dk/scheduling/schedulingfrontend/sharedcomponents/TimePicker.kt:47]:
```suggestion
```

**Comment by henneboy** [comment]
Created issues #60 and #61

File: frontend/app/src/main/java/dk/scheduling/schedulingfrontend/components/TimePicker.kt [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.remember
+import androidx.compose.ui.Alignment
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.tooling.preview.Preview
+import androidx.compose.ui.unit.dp
+import androidx.compose.ui.window.DialogProperties
+import dk.scheduling.schedulingfrontend.ui.theme.SchedulingFrontendTheme
+
+@OptIn(ExperimentalMaterial3Api::class)
```
Comment (Mikkel-Boje) [inline frontend/app/src/main/java/dk/scheduling/schedulingfrontend/components/TimePicker.kt:35]: I expect validation on all task's timespans on the frontend as well

**Comment by henneboy** [comment]
This branch should be squashed..

**Comment by henneboy** [comment]
> Kotlin imports 💀

Kun 128 linjer ud af 672 = 19%

File: frontend/app/src/main/java/dk/scheduling/schedulingfrontend/MainLayout.kt [hunk 8]
```diff
@@ -46,7 +46,7 @@ fun App() {
         NavHost(navController = navController, startDestination = Page.Home.route, modifier = Modifier.padding(innerPadding)) {
             composable(Page.Home.route) { HomePage(modifier = Modifier, getDevices = { testDeviceOverview() }) }
             composable(Page.ApiButton.route) { ApiButton() }
-            composable(Page.Page3.route) { Page3() }
+            composable(Page.Page3.route) { CreateTaskPage(Modifier, handleSubmission = {}, handleCancellation = {}) }
```
Comment (mads256h) [inline frontend/app/src/main/java/dk/scheduling/schedulingfrontend/MainLayout.kt:49]: Is this just for testing purposes?

File: frontend/app/src/main/java/dk/scheduling/schedulingfrontend/components/DateRangePicker.kt [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+        BasicAlertDialog(
+            onDismissRequest = {
+                // Dismiss the dialog when the user clicks outside the dialog or on the back
+                // button. If you want to disable that functionality, simply use an empty
+                // onDismissRequest.
+                closeDialog()
+            },
+            properties =
+                DialogProperties(
+                    usePlatformDefaultWidth = false,
```
Comment (mads256h) [inline frontend/app/src/main/java/dk/scheduling/schedulingfrontend/components/DateRangePicker.kt:51]: Hvorfor ikke default platform width? Er der for lidt plads i den?

File: frontend/app/src/main/java/dk/scheduling/schedulingfrontend/components/DateRangePicker.kt [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+                    // Add a row with "Save" and dismiss actions.
+                    Row(
+                        modifier =
+                            Modifier
+                                .fillMaxWidth(),
+                        verticalAlignment = Alignment.CenterVertically,
+                        horizontalArrangement = Arrangement.SpaceBetween,
+                    ) {
+                        IconButton(onClick = { closeDialog() }) {
+                            Icon(Icons.Filled.Close, contentDescription = "Localized description")
```
Comment (mads256h) [inline frontend/app/src/main/java/dk/scheduling/schedulingfrontend/components/DateRangePicker.kt:83]:
```suggestion
                            Icon(Icons.Filled.Close, contentDescription = "Close")
```

File: frontend/app/src/main/java/dk/scheduling/schedulingfrontend/components/StandardDropDownMenu.kt [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+import androidx.compose.runtime.remember
+import androidx.compose.runtime.setValue
+import androidx.compose.ui.Alignment
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.tooling.preview.Preview
+import androidx.compose.ui.unit.dp
+import dk.scheduling.schedulingfrontend.ui.theme.SchedulingFrontendTheme
+
+@Composable
+fun StandardDropDownMenu(
```
Comment (mads256h) [inline frontend/app/src/main/java/dk/scheduling/schedulingfrontend/components/StandardDropDownMenu.kt:29]: Er der ikke defineret en standard dropdown menu i jetpack compose?

File: frontend/app/src/main/java/dk/scheduling/schedulingfrontend/pages/CreateTaskPage.kt [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+        StandardDropDownMenu(
+            modifier = Modifier,
+            options = options,
+            label = "Devices",
+            selectedItem = selectedItem,
+            onSelect = { selectedItem = it },
+        )
+
+        // Duration input field
+        Spacer(modifier = Modifier.height(20.dp))
```
Comment (mads256h) [inline frontend/app/src/main/java/dk/scheduling/schedulingfrontend/pages/CreateTaskPage.kt:75]: Inconsistent spacer heights.
Can we use something other than spacers for everything?
Padding works better.

File: frontend/app/src/main/java/dk/scheduling/schedulingfrontend/pages/CreateTaskPage.kt [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+            onClick = {
+                handleCancellation()
+            },
+            text = "Cancel",
+        )
+    }
+}
+
+@Composable
+fun ClickableCard(
```
Comment (mads256h) [inline frontend/app/src/main/java/dk/scheduling/schedulingfrontend/pages/CreateTaskPage.kt:156]: It seems a little strange for this to be cards. Why not buttons?

File: frontend/app/src/test/java/dk/scheduling/schedulingfrontend/CreateTaskPageTest.kt [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+import dk.scheduling.schedulingfrontend.pages.isValidInput
+import org.junit.Test
+
+@OptIn(ExperimentalMaterial3Api::class)
+class CreateTaskPageTest {
+    private val validTime = TimePickerState(0, 0, true)
+    private val validTimeMidday = TimePickerState(12, 0, true)
+
+    @Test
+    fun isValidDurationTest() {
```
Comment (mads256h) [inline frontend/app/src/test/java/dk/scheduling/schedulingfrontend/CreateTaskPageTest.kt:16]: Cant we use a theory with datapoints?

**Review by mads256h — CHANGES_REQUESTED** [review]
Small stuff and questions

File: frontend/app/src/main/java/dk/scheduling/schedulingfrontend/MainLayout.kt (hunk 8 above)
Comment (henneboy) [reply frontend/app/src/main/java/dk/scheduling/schedulingfrontend/MainLayout.kt:49]: API integration will be in another PR

File: frontend/app/src/main/java/dk/scheduling/schedulingfrontend/components/DateRangePicker.kt (hunk 9 above)
Comment (henneboy) [reply frontend/app/src/main/java/dk/scheduling/schedulingfrontend/components/DateRangePicker.kt:51]: Nej, så stretcher den til at fylde hele skærmen men på en scuffed måde

File: frontend/app/src/test/java/dk/scheduling/schedulingfrontend/CreateTaskPageTest.kt (hunk 14 above)
Comment (henneboy) [reply frontend/app/src/test/java/dk/scheduling/schedulingfrontend/CreateTaskPageTest.kt:16]: No, then we would need to install some new package, which I dont know how to do :)

File: frontend/app/src/main/java/dk/scheduling/schedulingfrontend/pages/CreateTaskPage.kt (hunk 13 above)
Comment (henneboy) [reply frontend/app/src/main/java/dk/scheduling/schedulingfrontend/pages/CreateTaskPage.kt:156]: So it seems more like fields and less like buttons? We usually dont print info in the field of a button?

File: frontend/app/src/main/java/dk/scheduling/schedulingfrontend/pages/CreateTaskPage.kt (hunk 12 above)
Comment (henneboy) [reply frontend/app/src/main/java/dk/scheduling/schedulingfrontend/pages/CreateTaskPage.kt:75]: Padding is also applied on the sides, which we dont want, also due to the way the title is created we need a spacer in the top anyway

Comment (mads256h) [reply frontend/app/src/main/java/dk/scheduling/schedulingfrontend/pages/CreateTaskPage.kt:75]: You can specify which direction to pad with e.g. `.padding(bottom = 20.dp)`

File: frontend/app/src/main/java/dk/scheduling/schedulingfrontend/components/StandardDropDownMenu.kt (hunk 11 above)
Comment (henneboy) [reply frontend/app/src/main/java/dk/scheduling/schedulingfrontend/components/StandardDropDownMenu.kt:29]: Jo, det er også den vi bruger? https://developer.android.com/reference/kotlin/androidx/compose/material3/package-summary#ExposedDropdownMenuBox(kotlin.Boolean,kotlin.Function1,androidx.compose.ui.Modifier,kotlin.Function1)

File: frontend/app/src/main/java/dk/scheduling/schedulingfrontend/pages/CreateTaskPage.kt (hunk 12 above)
Comment (henneboy) [reply frontend/app/src/main/java/dk/scheduling/schedulingfrontend/pages/CreateTaskPage.kt]: @mads256h no more spacers :)

**Review by S3JER — APPROVED** [review]

**Review by Puvikaran2001 — APPROVED** [review]
Nice

**Pull request merged**